### PR TITLE
BUPH-145 | Shared Templates

### PR DESCRIPTION
**Pull request by jpmarcotte**
The important stuff is probably in the [2nd commit](07d6fef447aae3a0644a12464c16bda95a3e4198).

Definitely looking for feedback on the template strings that get turned into tokens and their usability. I feel like they could be better, but I haven't thought of anything just yet.

### CODE REVIEW

**Review by mucha55 — APPROVED** [review]

File: src/V2/Actor.php [hunk 1]
```diff
@@ -10,7 +10,10 @@ class Actor implements ActorInterface
     protected $NamespacePrefix;
     protected $NamespaceRelative;
     protected $RelativeClassPath;
-    protected $ParentRelativeClassPath;
+    protected $ParentActorRelativeClassPath;
```
Comment (arielallon) [inline src/V2/Actor.php:13]: probably a not-for-now, but is v2 an opportunity to force typed properties?

File: src/V2/Actor/Template/Tokenizer.php [hunk 2]
```diff
@@ -20,6 +20,10 @@ class Tokenizer implements TokenizerInterface
         '\RelativeNamespace' => self::TOKEN_NAMESPACE_RELATIVE,
         'NamespacedPrimaryActorName' => self::TOKEN_PRIMARY_ACTOR_NAME_FULL_PASCAL,
         'PrimaryActorName' => self::TOKEN_PRIMARY_ACTOR_NAME_SHORT_PASCAL,
+        'NamespacedParentActorName' => self::TOKEN_PARENT_ACTOR_NAME_FULL_PASCAL,
+        'RelativeParentActorName' => self::TOKEN_PARENT_ACTOR_NAME_RELATIVE_PASCAL,
+        'ParentActorName' => self::TOKEN_PARENT_ACTOR_NAME_SHORT_PASCAL,
```
[the remark this answers is not in the document]
Comment (arielallon) [reply src/V2/Actor/Template/Tokenizer.php:25]: or maybe at least leave a comment in the code here as to the ordering being important and what the reasoning is

**Review by arielallon — APPROVED** [review]
neat. this looks solid to me.

File: src/V2/Actor.php (hunk 1 above)
Comment (jpmarcotte) [reply src/V2/Actor.php:13]: Agreed on both counts. There's a whole lot of internal cleanup that can/should happen, but this is currently more focused on the changes for the API / functionality improvements.

File: src/V2/Actor/Template/Tokenizer.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpmarcotte) [reply src/V2/Actor/Template/Tokenizer.php:25]: Yeah, I think any time there's a bunch of search/replace things happening, order should always be taken into account. There's just too many edge cases to deal with otherwise. I'll add comments so they aren't as likely to get moved, but that's also why we have tests.

An alternative (not sure how well it would work for the API) is to use something like `ShortPrimaryActorName` and `ShortParentActorName` to prevent substrings.

**Comment by arielallon** [comment]
Looking through the available tokens, I can currently think of other things that we might want.

File: src/V2/Actor.php [hunk 3]
```diff
@@ -10,7 +10,10 @@ class Actor implements ActorInterface
...
+    protected $ParentActorFullPascalCaseName;
+    protected $ParentActorRelativePascalCaseName;
```
[the remark this answers is not in the document]
Comment (jpmarcotte) [reply src/V2/Actor.php:15]: These are both referenced in the `SharedActorName/AwareTrait.php` file.

File: tests/v2/SharedTemplates/templates/PrimaryActorName/Map.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        PrimaryActorNameInterface ...$PrimaryActorNames
+    ): MapInterface {
+        return $this;
+    }
+
+    #[\ReturnTypeWillChange]
+    public function getArrayCopy(): MapInterface
+    {
+        return new self(parent::getArrayCopy(), (int)$this->getFlags());
+    }
```
[the remark this answers is not in the document]
Comment (jpmarcotte) [reply tests/v2/SharedTemplates/templates/PrimaryActorName/Map.php:65]: https://php.watch/versions/8.1/ReturnTypeWillChange

It's future-proofing for supporting PHP 8.1.

File: tests/v2/SharedTemplates/control/DOR/Version/Car/Map/Repository.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?php
```
[the remark this answers is not in the document]
Comment (jpmarcotte) [reply tests/v2/SharedTemplates/control/DOR/Version/Car/Map/Repository.php:1]: This is mostly a Copy/Paste thing to show that the same stuff can be created as we had in the previous "full test" that was the Buphalo Fitness Demo.

File: tests/v2/SharedTemplates/templates/PrimaryActorName/Repository.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+use Neighborhoods\BuphaloTemplateTree\RelativeNamespace\PrimaryActorName;
+use Neighborhoods\BuphaloTemplateTree\RelativeNamespace\PrimaryActorNameInterface;
+
+class Repository implements RepositoryInterface
+{
+    use PrimaryActorName\Builder\Factory\AwareTrait;
+
+    public function add(PrimaryActorNameInterface $PrimaryActorName): RepositoryInterface
+    {
+        /** @neighborhoods-buphalo:annotation-processor \Neighborhoods\BuphaloTemplateTree\RelativeNamespace\PrimaryActorName\Repository.add1
```
[the remark this answers is not in the document]
Comment (jpmarcotte) [reply tests/v2/SharedTemplates/templates/PrimaryActorName/Repository.php:15]: I don't think it's necessary so much as a copy/paste holdover from the early days. As far as I can remember, there isn't any necessary uniqueness as annotations are only done per-template, and even then, they don't have to be unique, but if they aren't, then they'll get replaced by the same thing (which is sometimes desirable).

It's possible it was put there as a future-proof in case things like global-scope annotation tags were introduced.

**Comment by jpmarcotte** [comment]
> I can currently think of other things that we might want.

Such as?

Also, eventually we'll swap these out so they can be more customizable, but I felt like supporting shared templates as a feature was more important.

**Comment by arielallon** [comment]
>> I can currently think of other things that we might want.
>
> Such as?

hahaha woops, that should have said "can't".
